File: Paw_Track-Anshu_IN26011396/src/services/adoptionService.ts
import { dbContext } from '../data/dbContext';
import { AdoptionApplication, User } from '../data/schema';

export interface SubmitApplicationDto {
  animalId: number;
  visitBookingId?: number | null;
  notes?: string;
}

export class AdoptionService {
  /**
   * Submit an adoption application for an animal.
   * Validates:
   * Rule 1 — Animal must be "Available"
   * Rule 2 — Adopter cannot have an active (Pending) application for the same animal
   * Rule 3 — If VisitBookingId is provided:
   *          - VisitBooking must exist
   *          - VisitBooking.AdopterId == current adopter
   *          - VisitBooking.AnimalId == application AnimalId
   */
  public submitApplication(dto: SubmitApplicationDto, currentUser: User): AdoptionApplication {
    if (currentUser.role !== 'Adopter' && currentUser.role !== 'OrgAdmin') {
      // Adopters submit applications for themselves
    }

    const adopterId = currentUser.id;

    // 1. Fetch Animal & Validate Availability
    const animal = dbContext.animals.find((a) => a.id === dto.animalId);
    if (!animal) {
      throw new Error('Animal not found.');
    }

    if (animal.status !== 'Available') {
      throw new Error(`Animal is no longer available. Current status: ${animal.status}.`);
    }

    // 2. Prevent duplicate active application
    const existingActiveApp = dbContext.adoptionApplications.find(
      (app) => app.adopterId === adopterId && app.animalId === dto.animalId && app.status === 'Pending'
    );

    if (existingActiveApp) {
      throw new Error('An active application already exists for this adopter and animal.');
    }

    // 3. Validate VisitBooking integration if supplied
    if (dto.visitBookingId) {
      const visit = dbContext.visitBookings.find((v) => v.id === dto.visitBookingId);
      if (!visit) {
        throw new Error(`Visit booking with ID ${dto.visitBookingId} was not found.`);
      }

      if (visit.adopterId !== adopterId) {
        throw new Error('The selected visit does not belong to this adopter.');
      }

      if (visit.animalId !== dto.animalId) {
        throw new Error('The selected visit belongs to a different animal.');
      }
    }

    // Create Application
    const newApplication: AdoptionApplication = {
      id: dbContext.getNextId('adoptionApplications'),
      animalId: dto.animalId,
      adopterId: adopterId,
      visitBookingId: dto.visitBookingId || null,
      applicationDate: new Date().toISOString().split('T')[0],
      status: 'Pending',
      notes: dto.notes || ''
    };

    dbContext.adoptionApplications.push(newApplication);
    return newApplication;
  }

  /**
   * Get applications filtered by branch authorization and status
   */
  public getApplications(currentUser: User, statusFilter?: string, branchFilter?: number): any[] {
    let apps = [...dbContext.adoptionApplications];

    // Branch authorization scoping
    if (currentUser.role !== 'OrgAdmin' && currentUser.branchId !== null) {
      // Branch scoped staff only see applications for animals in their branch
      const branchAnimalIds = new Set(
        dbContext.animals.filter((a) => a.branchId === currentUser.branchId).map((a) => a.id)
      );
      apps = apps.filter((app) => branchAnimalIds.has(app.animalId));
    } else if (branchFilter) {
      const branchAnimalIds = new Set(
        dbContext.animals.filter((a) => a.branchId === branchFilter).map((a) => a.id)
      );
      apps = apps.filter((app) => branchAnimalIds.has(app.animalId));
    }

    if (statusFilter && statusFilter !== 'All') {
      apps = apps.filter((app) => app.status === statusFilter);
    }

    // Enrich with detailed information for Presentation layer
    return apps.map((app) => {
      const animal = dbContext.animals.find((a) => a.id === app.animalId);
      const adopter = dbContext.users.find((u) => u.id === app.adopterId);
      const branch = animal ? dbContext.branches.find((b) => b.id === animal.branchId) : null;
      const visit = app.visitBookingId
        ? dbContext.visitBookings.find((v) => v.id === app.visitBookingId)
        : null;
      const visitSlot = visit
        ? dbContext.visitSlots.find((s) => s.id === visit.visitSlotId)
        : null;

      return {
        ...app,
        animalName: animal?.name || 'Unknown',
        animalSpecies: animal?.species || '',
        animalBreed: animal?.breed || '',
        animalStatus: animal?.status || '',
        adopterName: adopter?.name || 'Unknown',
        adopterEmail: adopter?.email || '',
        branchId: animal?.branchId,
        branchName: branch?.name || 'Unknown',
        linkedVisit: visit
          ? {
              id: visit.id,
              bookingDate: visit.bookingDate,
              status: visit.status,
              slotDate: visitSlot?.date,
              slotTime: visitSlot ? `${visitSlot.startTime} - ${visitSlot.endTime}` : ''
            }
          : null
      };
    });
  }

  /**
   * Get application by ID with branch security check
   */
  public getApplicationById(id: number, currentUser: User): any {
    const apps = this.getApplications(currentUser);
    const app = apps.find((a) => a.id === id);
    if (!app) {
      throw new Error(`Application with ID ${id} not found or access denied.`);
    }
    return app;
  }
}

export const adoptionService = new AdoptionService();

File: Paw_Track-Anshu_IN26011396/src/PawTrack.Services/AdoptionService.cs
using Microsoft.EntityFrameworkCore;
using PawTrack.Core.DTOs;
using PawTrack.Core.Entities;
using PawTrack.Core.Enums;
using PawTrack.Core.Interfaces;
using PawTrack.Data;

namespace PawTrack.Services;

public class AdoptionService : IAdoptionService
{
    private readonly PawTrackDbContext _context;

    public AdoptionService(PawTrackDbContext context)
    {
        _context = context;
    }

    /// <summary>
    /// Submit an adoption application for an animal.
    /// Validates:
    /// Rule 1 — Animal must be "Available"
    /// Rule 2 — Adopter cannot have an active (Pending) application for the same animal
    /// Rule 3 — If VisitBookingId is provided:
    ///          - VisitBooking must exist
    ///          - VisitBooking.AdopterId == current adopter
    ///          - VisitBooking.AnimalId == application AnimalId
    /// </summary>
    public async Task<AdoptionApplication> SubmitApplicationAsync(SubmitApplicationDto dto, User currentUser)
    {
        var adopterId = currentUser.Id;

        // 1. Fetch Animal & Validate Availability
        var animal = await _context.Animals.FirstOrDefaultAsync(a => a.Id == dto.AnimalId);
        if (animal == null)
        {
            throw new KeyNotFoundException("Animal not found.");
        }

        if (animal.Status != AnimalStatus.Available)
        {
            throw new InvalidOperationException($"Animal is no longer available. Current status: {animal.Status}.");
        }

        // 2. Prevent duplicate active (Pending) application for the same adopter and animal
        var existingActiveApp = await _context.AdoptionApplications
            .AnyAsync(app => app.AdopterId == adopterId && app.AnimalId == dto.AnimalId && app.Status == ApplicationStatus.Pending);

        if (existingActiveApp)
        {
            throw new InvalidOperationException("An active application already exists for this adopter and animal.");
        }

        // 3. Validate VisitBooking integration if supplied
        if (dto.VisitBookingId.HasValue && dto.VisitBookingId.Value > 0)
        {
            var visit = await _context.VisitBookings.FirstOrDefaultAsync(v => v.Id == dto.VisitBookingId.Value);
            if (visit == null)
            {
                throw new KeyNotFoundException($"Visit booking with ID {dto.VisitBookingId.Value} was not found.");
            }

            if (visit.AdopterId != adopterId)
            {
                throw new InvalidOperationException("The selected visit does not belong to this adopter.");
            }

            if (visit.AnimalId != dto.AnimalId)
            {
                throw new InvalidOperationException("The selected visit belongs to a different animal.");
            }
        }

        var newApplication = new AdoptionApplication
        {
            AnimalId = dto.AnimalId,
            AdopterId = adopterId,
            VisitBookingId = dto.VisitBookingId > 0 ? dto.VisitBookingId : null,
            ApplicationDate = DateTime.UtcNow.Date,
            Status = ApplicationStatus.Pending,
            Notes = dto.Notes ?? string.Empty
        };

        _context.AdoptionApplications.Add(newApplication);
        await _context.SaveChangesAsync();

        return newApplication;
    }

    /// <summary>
    /// Get applications filtered by branch authorization and status
    /// </summary>
    public async Task<List<ApplicationResponseDto>> GetApplicationsAsync(User currentUser, ApplicationStatus? statusFilter = null, int? branchFilter = null)
    {
        var query = _context.AdoptionApplications
            .Include(app => app.Animal)
                .ThenInclude(an => an!.Branch)
            .Include(app => app.Adopter)
            .Include(app => app.VisitBooking)
                .ThenInclude(v => v!.VisitSlot)
            .AsQueryable();

        // Branch authorization scoping
        if (currentUser.Role == UserRole.Adopter)
        {
            query = query.Where(app => app.AdopterId == currentUser.Id);
        }
        else if (currentUser.Role != UserRole.OrgAdmin && currentUser.BranchId.HasValue)
        {
            query = query.Where(app => app.Animal != null && app.Animal.BranchId == currentUser.BranchId.Value);
        }
        else if (branchFilter.HasValue && branchFilter.Value > 0)
        {
            query = query.Where(app => app.Animal != null && app.Animal.BranchId == branchFilter.Value);
        }

        if (statusFilter.HasValue)
        {
            query = query.Where(app => app.Status == statusFilter.Value);
        }

        var apps = await query.OrderByDescending(app => app.ApplicationDate).ToListAsync();

        return apps.Select(app => new ApplicationResponseDto
        {
            Id = app.Id,
            AnimalId = app.AnimalId,
            AnimalName = app.Animal?.Name ?? "Unknown",
            AnimalSpecies = app.Animal?.Species ?? string.Empty,
            AnimalBreed = app.Animal?.Breed ?? string.Empty,
            AnimalStatus = app.Animal?.Status ?? AnimalStatus.UnderAssessment,
            AdopterId = app.AdopterId,
            AdopterName = app.Adopter?.Name ?? "Unknown",
            AdopterEmail = app.Adopter?.Email ?? string.Empty,
            BranchId = app.Animal?.BranchId,
            BranchName = app.Animal?.Branch?.Name ?? "Unknown",
            VisitBookingId = app.VisitBookingId,
            ApplicationDate = app.ApplicationDate,
            Status = app.Status,
            RejectionReason = app.RejectionReason,
            Notes = app.Notes,
            LinkedVisit = app.VisitBooking != null ? new LinkedVisitDto
            {
                Id = app.VisitBooking.Id,
                BookingDate = app.VisitBooking.BookingDate,
                Status = app.VisitBooking.Status,
                SlotDate = app.VisitBooking.VisitSlot?.Date,
                SlotTime = app.VisitBooking.VisitSlot != null
                    ? $"{app.VisitBooking.VisitSlot.StartTime:hh\\:mm} - {app.VisitBooking.VisitSlot.EndTime:hh\\:mm}"
                    : string.Empty
            } : null
        }).ToList();
    }

    /// <summary>
    /// Get application by ID with branch security check
    /// </summary>
    public async Task<ApplicationResponseDto> GetApplicationByIdAsync(int id, User currentUser)
    {
        var apps = await GetApplicationsAsync(currentUser);
        var app = apps.FirstOrDefault(a => a.Id == id);
        if (app == null)
        {
            throw new KeyNotFoundException($"Application with ID {id} not found or access denied.");
        }
        return app;
    }
}
